tests/trinity/core/p2p-proto/test_node_data_request_object.py

```diff
@@ -0,0 +1,74 @@
+import os
+import random
+
+import pytest
+
+from eth_utils import keccak
+
+from p2p.exceptions import ValidationError
+
+from trinity.protocol.eth.requests import NodeDataRequest
+
+
+def mk_node():
+    node_length = random.randint(0, 2048)
+    node = os.urandom(node_length)
+    return node
+
+
+def mk_node_data(n):
+    if n == 0:
+        return tuple(), tuple()
+    nodes = tuple(set(mk_node() for _ in range(n)))
+    node_keys = tuple(keccak(node) for node in nodes)
+    return node_keys, nodes
+
+
+def test_node_data_request_empty_response_is_valid():
+    node_keys, _ = mk_node_data(10)
+    request = NodeDataRequest(node_keys)
+
+    request.validate_response(tuple())
+
+
+def test_node_data_request_with_full_response():
+    node_keys, nodes = mk_node_data(10)
+    request = NodeDataRequest(node_keys)
+    node_data = tuple(zip(node_keys, nodes))
+
+    request.validate_response(node_data)
+
+
+def test_node_data_request_with_partial_response():
+    node_keys, nodes = mk_node_data(10)
+    request = NodeDataRequest(node_keys)
+    node_data = tuple(zip(node_keys, nodes))
+
+    request.validate_response(node_data[3:])
+    request.validate_response(node_data[:3])
+    request.validate_response((node_data[1], node_data[8], node_data[4]))
+
+
+def test_node_data_request_with_fully_invalid_response():
+    node_keys, nodes = mk_node_data(10)
+    request = NodeDataRequest(node_keys)
+
+    # construct a unique set of other nodes
+    other_nodes = tuple(set(mk_node() for _ in range(10)).difference(nodes))
+    other_node_data = tuple((keccak(node), node) for node in other_nodes)
+
+    with pytest.raises(ValidationError):
+        request.validate_response(other_node_data)
+
+
+def test_node_data_request_with_extra_unrequested_nodes():
+    node_keys, nodes = mk_node_data(10)
+    request = NodeDataRequest(node_keys)
+    node_data = tuple(zip(node_keys, nodes))
+
+    # construct a unique set of other nodes
+    other_nodes = tuple(set(mk_node() for _ in range(10)).difference(nodes))
+    other_node_data = tuple((keccak(node), node) for node in other_nodes)
+
+    with pytest.raises(ValidationError):
```

Comment:
I think `ValidationError` is the semantically correct one here.

- `MalformedMessage`: The data payload sent is not well formed and thus cannot be decoded/parsed/etc.
- `ValidationError`: The data is well formed, but the actual content of it does not pass some validation check (like that they sent us the stuff we asked for).

Also, this exception doesn't escape the Manager's internal message loop and is used a signal from `Request.validate_response` to indicate that the response does not validate and thus should not fulfil the request.

That does however point out that the `_normalize_response`  API should probably be raising `MalformedMessage` exceptions since the checks it performs are about whether the data is well formed.  Will fix.